core/sile.lua

```diff
@@ -240,10 +240,59 @@ function SILE.resolveFile(filename, pathprefix)
   return nil
 end
 
+SILE.currentCommandStack = {}
+--- Push currently processed command on to the stack to record the execution trace for debugging.
+function SILE.currentCommandStack.push(stack, command)
+  if command == nil then return end
+
+  stack[#stack + 1] = command
+  SILE.currentCommand = command
+end
+--- Pop previously pushed command from the stack.
+--- Command is provided to check that pushes and pops are balanced.
+function SILE.currentCommandStack.pop(stack, command)
+  if command == nil then return end
+
+  local popped = stack[#stack]
+  if popped ~= command then
+    SU.warn("Unbalanced command push/pop. Expected '" .. popped .. "', got '" .. command .. "'.")
```

Comment:
See overall review notes, but this fault condition will be reached in a lot of documents because it doesn't allow for there being multiple typesetters.

Reply:
Can you please give an example, where multiple typesetters are used in parallel?

Reply:
Checkout packages such as footnotes: https://github.com/simoncozens/sile/blob/c6a0c07911e50448acea660fcaf9a7d1a8bdaa55/packages/footnotes.lua#L40-L53 or [parallel](https://github.com/simoncozens/sile/blob/master/packages/parallel.lua).